src/urllib3/_collections.py

```diff
@@ -28,10 +55,33 @@ def __exit__(self, exc_type, exc_value, traceback):
 __all__ = ["RecentlyUsedContainer", "HTTPHeaderDict"]
 
 
-_Null = object()
+_KT = TypeVar("_KT")
+_VT = TypeVar("_VT")
 
+ValidHttpHeaderSource = Union[
+    "HTTPHeaderDict",
+    Mapping[str, str],
+    Iterable[Tuple[str, str]],
+    "HasGettableStringKeys",
+]
 
-class RecentlyUsedContainer(MutableMapping):
+
+def ensure_can_construct_http_header_dict(
+    potential: object,
+) -> Optional[ValidHttpHeaderSource]:
+    if isinstance(potential, HTTPHeaderDict):
+        return potential
+    elif isinstance(potential, Mapping):
```

Comment:
If we have a `Mapping` or an `Iterable`, we pretty much assume that the library passing it to us knew what they were doing and made it the right shape.